File: tsrs2/crates/oracle/identity.mjs
// A2 exact scope identity — Node half of the versioned canonical
// occurrence encoder (measurement-integrity.md §3, encoder v1).
//
// Mirrors crates/conformance/src/identity.rs byte for byte. The scope
// audit feeds golden oracle records and the committed vector file
// through both encoders and fails on any difference. This tool is
// check-only and NOT part of the golden-producing path (driver.mjs /
// program-host.mjs); it must never be imported from the pinned
// producer modules, or the A1 producer pin would go stale.
//
// stdin:  {"encoder":1,"cases":[{"fixture","matrix_key","records":[...]}]}
// stdout: {"encoder":1,"cases":[{"record_canonical":[...],
//          "identities":[...],"identity_canonical":[...],
//          "identity_sha256":[...]}]}

import { createHash } from "node:crypto";

const ENCODER_VERSION = 1;

// JSON string escaping, matching the Rust writer exactly: `"`, `\`,
// shorthands \b \t \n \f \r, other control chars below 0x20 as
// lowercase \u00xx, everything else raw UTF-8. Deliberately explicit
// instead of JSON.stringify so both languages pin the same table.
function escapeString(value) {
  let out = '"';
  for (const ch of value) {
    const point = ch.codePointAt(0);
    if (ch === '"') out += '\\"';
    else if (ch === "\\") out += "\\\\";
    else if (point === 0x08) out += "\\b";
    else if (point === 0x09) out += "\\t";
    else if (point === 0x0a) out += "\\n";
    else if (point === 0x0c) out += "\\f";
    else if (point === 0x0d) out += "\\r";
    else if (point < 0x20) out += "\\u" + point.toString(16).padStart(4, "0");
    else out += ch;
  }
  return out + '"';
}

function encodeOptString(value) {
  return value === null || value === undefined ? "null" : escapeString(value);
}

function encodeOptInt(value) {
  if (value === null || value === undefined) return "null";
  if (!Number.isInteger(value) || value < 0) {
    throw new Error(`canonical integers are unsigned decimals, got ${value}`);
  }
  return String(value);
}

function encodeChain(chain) {
  const next = (chain.next ?? []).map(encodeChain).join(",");
  return (
    `{"category":${escapeString(chain.category)}` +
    `,"code":${encodeOptInt(chain.code)}` +
    `,"next":[${next}]` +
    `,"text":${escapeString(chain.text)}}`
  );
}

function encodeRelated(related) {
  const entries = (related ?? []).map(
    (entry) =>
      `{"category":${escapeString(entry.category)}` +
      `,"chain":${encodeChain(entry.chain)}` +
      `,"code":${encodeOptInt(entry.code)}` +
      `,"file":${encodeOptString(entry.file)}` +
      `,"length":${encodeOptInt(entry.length)}` +
      `,"start":${encodeOptInt(entry.start)}}`,
  );
  return `[${entries.join(",")}]`;
}

function encodeRecord(diag) {
  return (
    `{"category":${escapeString(diag.category)}` +
    `,"chain":${encodeChain(diag.chain)}` +
    `,"code":${encodeOptInt(diag.code)}` +
    `,"col":${encodeOptInt(diag.col)}` +
    `,"file":${encodeOptString(diag.file)}` +
    `,"length":${encodeOptInt(diag.length)}` +
    `,"line":${encodeOptInt(diag.line)}` +
    `,"pass":${encodeOptString(diag.pass)}` +
    `,"related":${encodeRelated(diag.related)}` +
    `,"reports_deprecated":${diag.reports_deprecated ? "true" : "false"}` +
    `,"reports_unnecessary":${diag.reports_unnecessary ? "true" : "false"}` +
    `,"source":${encodeOptString(diag.source)}` +
    `,"start":${encodeOptInt(diag.start)}}`
  );
}

function encodeIdentity(identity) {
  return (
    `{"category":${escapeString(identity.category)}` +
    `,"chain_sha256":${escapeString(identity.chain_sha256)}` +
    `,"code":${encodeOptInt(identity.code)}` +
    `,"file":${encodeOptString(identity.file)}` +
    `,"fixture":${escapeString(identity.fixture)}` +
    `,"length":${encodeOptInt(identity.length)}` +
    `,"matrix_key":${escapeString(identity.matrix_key)}` +
    `,"occurrence":${encodeOptInt(identity.occurrence)}` +
    `,"pass":${escapeString(identity.pass)}` +
    `,"related_sha256":${escapeString(identity.related_sha256)}` +
    `,"start":${encodeOptInt(identity.start)}}`
  );
}

function sha256Hex(text) {
  return createHash("sha256").update(Buffer.from(text, "utf8")).digest("hex");
}

function caseReport(fixture, matrixKey, records) {
  const identities = records.map((diag) => {
    if (diag.pass === null || diag.pass === undefined) {
      throw new Error(
        `golden ${fixture} [${matrixKey}] lacks pass provenance for code ${diag.code}`,
      );
    }
    return {
      fixture,
      matrix_key: matrixKey,
      pass: diag.pass,
      file: diag.file ?? null,
      start: diag.start ?? null,
      length: diag.length ?? null,
      code: diag.code,
      category: diag.category,
      chain_sha256: sha256Hex(encodeChain(diag.chain)),
      related_sha256: sha256Hex(encodeRelated(diag.related)),
      occurrence: 0,
    };
  });

  // Stable sort by complete canonical record BYTES (UTF-8 order, not
  // UTF-16 code-unit order), byte-identical neighbors retaining input
  // order; then number occurrences per identity tuple in that order.
  const recordCanonical = records.map(encodeRecord);
  const recordBytes = recordCanonical.map((text) => Buffer.from(text, "utf8"));
  const order = recordBytes
    .map((_, index) => index)
    .sort((a, b) => Buffer.compare(recordBytes[a], recordBytes[b]) || a - b);
  const counts = new Map();
  for (const index of order) {
    const tupleKey = encodeIdentity(identities[index]); // occurrence still 0
    const seen = counts.get(tupleKey) ?? 0;
    identities[index].occurrence = seen;
    counts.set(tupleKey, seen + 1);
  }

  return {
    record_canonical: recordCanonical,
    identities,
    identity_canonical: identities.map(encodeIdentity),
    identity_sha256: identities.map((identity) =>
      sha256Hex(encodeIdentity(identity)),
    ),
  };
}

async function main() {
  const chunks = [];
  for await (const chunk of process.stdin) chunks.push(chunk);
  const input = JSON.parse(Buffer.concat(chunks).toString("utf8"));
  if (input.encoder !== ENCODER_VERSION) {
    throw new Error(
      `unsupported encoder version ${input.encoder} (this tool implements v${ENCODER_VERSION})`,
    );
  }
  const cases = input.cases.map((entry) =>
    caseReport(entry.fixture, entry.matrix_key, entry.records),
  );
  process.stdout.write(JSON.stringify({ encoder: ENCODER_VERSION, cases }));
}

main().catch((error) => {
  console.error(String(error?.stack ?? error));
  process.exit(1);
});

File: crates/conformance/src/identity.rs
//! A2 exact scope identity — the versioned canonical occurrence
//! encoder (measurement-integrity.md §3).
//!
//! Schema 2 identifies one oracle diagnostic occurrence as
//! `fixture + matrix_key + pass + file + start + length + code +
//! category + message-chain hash + related-information hash +
//! occurrence`. Rust and Node share this encoder version:
//! `crates/oracle/identity.mjs` mirrors every byte rule below and the
//! scope audit compares the two outputs over the committed vector file
//! and the corpus duplicate-bucket canaries. Changing any byte rule is
//! A2's one reviewed `input-schema-extension`, never a silent edit —
//! bump [`ENCODER_VERSION`] there and only there.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use super::{ConformanceResult, GoldenDiag, GoldenMessageChain, GoldenRelated};
use crate::ratchet::sha256_hex;

/// Version of the canonical byte encoding shared with
/// `crates/oracle/identity.mjs`. The scope manifest pins it.
pub const ENCODER_VERSION: u32 = 1;

/// One exact oracle diagnostic occurrence (measurement-integrity.md
/// §3). Line and column are deliberately absent: they are redundant
/// review fields on the manifest entry, verified against `start`,
/// never identity.
#[derive(Clone, Debug, Deserialize, Eq, Ord, PartialEq, PartialOrd, Serialize)]
#[serde(deny_unknown_fields)]
pub struct ExactIdentity {
    pub fixture: String,
    pub matrix_key: String,
    pub pass: String,
    pub file: Option<String>,
    pub start: Option<u32>,
    pub length: Option<u32>,
    pub code: u32,
    pub category: String,
    pub chain_sha256: String,
    pub related_sha256: String,
    pub occurrence: u32,
}

impl ExactIdentity {
    /// Human-readable identity label for error messages.
    pub fn label(&self) -> String {
        format!(
            "{} [{}] {}/{:?}/{}#{}",
            self.fixture, self.matrix_key, self.pass, self.file, self.code, self.occurrence
        )
    }

    /// Canonical identity bytes (alphabetical field order, v1).
    pub fn canonical_bytes(&self) -> Vec<u8> {
        let mut out = Vec::new();
        out.push(b'{');
        write_key(&mut out, "category", false);
        write_string(&mut out, &self.category);
        write_key(&mut out, "chain_sha256", true);
        write_string(&mut out, &self.chain_sha256);
        write_key(&mut out, "code", true);
        write_u32(&mut out, self.code);
        write_key(&mut out, "file", true);
        write_opt_string(&mut out, self.file.as_deref());
        write_key(&mut out, "fixture", true);
        write_string(&mut out, &self.fixture);
        write_key(&mut out, "length", true);
        write_opt_u32(&mut out, self.length);
        write_key(&mut out, "matrix_key", true);
        write_string(&mut out, &self.matrix_key);
        write_key(&mut out, "occurrence", true);
        write_u32(&mut out, self.occurrence);
        write_key(&mut out, "pass", true);
        write_string(&mut out, &self.pass);
        write_key(&mut out, "related_sha256", true);
        write_string(&mut out, &self.related_sha256);
        write_key(&mut out, "start", true);
        write_opt_u32(&mut out, self.start);
        out.push(b'}');
        out
    }

    pub fn sha256(&self) -> String {
        sha256_hex(&self.canonical_bytes())
    }
}

// ---------------------------------------------------------------------------
// Canonical byte writer (encoder v1)
// ---------------------------------------------------------------------------
// UTF-8, fixed (alphabetical) object field order, decimal integers,
// JSON string escaping, no insignificant whitespace, observable array
// order, `null` for absent optionals. Missing and empty stay distinct
// (`null` vs `""`/`[]`). The escape table matches the JSON standard's
// minimal set exactly — `"`, `\`, and control characters below 0x20
// (shorthand for \b \t \n \f \r, lowercase `\u00xx` otherwise);
// everything else is raw UTF-8. `identity.mjs` implements the same
// table explicitly rather than trusting `JSON.stringify`.

fn write_key(out: &mut Vec<u8>, key: &str, comma: bool) {
    if comma {
        out.push(b',');
    }
    write_string(out, key);
    out.push(b':');
}

fn write_string(out: &mut Vec<u8>, value: &str) {
    out.push(b'"');
    for ch in value.chars() {
        match ch {
            '"' => out.extend_from_slice(b"\\\""),
            '\\' => out.extend_from_slice(b"\\\\"),
            '\u{08}' => out.extend_from_slice(b"\\b"),
            '\t' => out.extend_from_slice(b"\\t"),
            '\n' => out.extend_from_slice(b"\\n"),
            '\u{0c}' => out.extend_from_slice(b"\\f"),
            '\r' => out.extend_from_slice(b"\\r"),
            ch if (ch as u32) < 0x20 => {
                out.extend_from_slice(format!("\\u{:04x}", ch as u32).as_bytes());
            }
            ch => {
                let mut buf = [0u8; 4];
                out.extend_from_slice(ch.encode_utf8(&mut buf).as_bytes());
            }
        }
    }
    out.push(b'"');
}

fn write_u32(out: &mut Vec<u8>, value: u32) {
    out.extend_from_slice(value.to_string().as_bytes());
}

fn write_bool(out: &mut Vec<u8>, value: bool) {
    out.extend_from_slice(if value { b"true".as_slice() } else { b"false" });
}

fn write_null(out: &mut Vec<u8>) {
    out.extend_from_slice(b"null");
}

fn write_opt_string(out: &mut Vec<u8>, value: Option<&str>) {
    match value {
        Some(value) => write_string(out, value),
        None => write_null(out),
    }
}

fn write_opt_u32(out: &mut Vec<u8>, value: Option<u32>) {
    match value {
        Some(value) => write_u32(out, value),
        None => write_null(out),
    }
}

/// Message-chain bytes: recursively text, code, category, children
/// (measurement-integrity.md §3).
fn write_chain(out: &mut Vec<u8>, chain: &GoldenMessageChain) {
    out.push(b'{');
    write_key(out, "category", false);
    write_string(out, &chain.category);
    write_key(out, "code", true);
    write_u32(out, chain.code);
    write_key(out, "next", true);
    out.push(b'[');
    for (index, child) in chain.next.iter().enumerate() {
        if index > 0 {
            out.push(b',');
        }
        write_chain(out, child);
    }
    out.push(b']');
    write_key(out, "text", true);
    write_string(out, &chain.text);
    out.push(b'}');
}

pub(crate) fn chain_bytes(chain: &GoldenMessageChain) -> Vec<u8> {
    let mut out = Vec::new();
    write_chain(&mut out, chain);
    out
}

/// Related-information bytes: the same diagnostic fields plus the
/// normalized virtual file/span, in observable array order. A reorder
/// of this array changes the identity.
pub(crate) fn related_bytes(related: &[GoldenRelated]) -> Vec<u8> {
    let mut out = Vec::new();
    out.push(b'[');
    for (index, entry) in related.iter().enumerate() {
        if index > 0 {
            out.push(b',');
        }
        out.push(b'{');
        write_key(&mut out, "category", false);
        write_string(&mut out, &entry.category);
        write_key(&mut out, "chain", true);
        write_chain(&mut out, &entry.chain);
        write_key(&mut out, "code", true);
        write_u32(&mut out, entry.code);
        write_key(&mut out, "file", true);
        write_opt_string(&mut out, entry.file.as_deref());
        write_key(&mut out, "length", true);
        write_opt_u32(&mut out, entry.length);
        write_key(&mut out, "start", true);
        write_opt_u32(&mut out, entry.start);
        out.push(b'}');
    }
    out.push(b']');
    out
}

/// Complete canonical record bytes (excluding occurrence, which is
/// assigned FROM this ordering). Every stored golden field appears so
/// two records differing anywhere order deterministically.
pub(crate) fn record_bytes(diag: &GoldenDiag) -> Vec<u8> {
    let mut out = Vec::new();
    out.push(b'{');
    write_key(&mut out, "category", false);
    write_string(&mut out, &diag.category);
    write_key(&mut out, "chain", true);
    write_chain(&mut out, &diag.chain);
    write_key(&mut out, "code", true);
    write_u32(&mut out, diag.code);
    write_key(&mut out, "col", true);
    write_opt_u32(&mut out, diag.col);
    write_key(&mut out, "file", true);
    write_opt_string(&mut out, diag.file.as_deref());
    write_key(&mut out, "length", true);
    write_opt_u32(&mut out, diag.length);
    write_key(&mut out, "line", true);
    write_opt_u32(&mut out, diag.line);
    write_key(&mut out, "pass", true);
    write_opt_string(&mut out, diag.pass.as_deref());
    write_key(&mut out, "related", true);
    out.extend_from_slice(&related_bytes(&diag.related));
    write_key(&mut out, "reports_deprecated", true);
    write_bool(&mut out, diag.reports_deprecated);
    write_key(&mut out, "reports_unnecessary", true);
    write_bool(&mut out, diag.reports_unnecessary);
    write_key(&mut out, "source", true);
    write_opt_string(&mut out, diag.source.as_deref());
    write_key(&mut out, "start", true);
    write_opt_u32(&mut out, diag.start);
    out.push(b'}');
    out
}

/// Assign exact identities to one case's oracle records, parallel to
/// the input order. `occurrence` is zero-based over records sharing
/// the identity tuple, numbered after stable sorting by complete
/// canonical record bytes — byte-identical neighbors retain oracle
/// input order, and same-tuple records that differ in a non-identity
/// field order deterministically by those bytes.
pub(crate) fn assign_case_identities(
    fixture: &str,
    matrix_key: &str,
    oracle: &[GoldenDiag],
) -> ConformanceResult<Vec<ExactIdentity>> {
    let mut tuples = Vec::with_capacity(oracle.len());
    for diag in oracle {
        let Some(pass) = diag.pass.as_deref() else {
            return Err(format!(
                "golden {fixture} [{matrix_key}] lacks pass provenance for code {}; exact \
                 scope identity requires schema-2 goldens (run `cargo xtask oracle-refresh`)",
                diag.code
            )
            .into());
        };
        tuples.push(ExactIdentity {
            fixture: fixture.to_owned(),
            matrix_key: matrix_key.to_owned(),
            pass: pass.to_owned(),
            file: diag.file.clone(),
            start: diag.start,
            length: diag.length,
            code: diag.code,
            category: diag.category.clone(),
            chain_sha256: sha256_hex(&chain_bytes(&diag.chain)),
            related_sha256: sha256_hex(&related_bytes(&diag.related)),
            occurrence: 0,
        });
    }

    let record_bytes = oracle.iter().map(record_bytes).collect::<Vec<_>>();
    let mut order = (0..oracle.len()).collect::<Vec<_>>();
    order.sort_by(|&a, &b| record_bytes[a].cmp(&record_bytes[b]).then(a.cmp(&b)));

    let mut counts: BTreeMap<ExactIdentity, u32> = BTreeMap::new();
    for index in order {
        let occurrence = {
            let counter = counts.entry(tuples[index].clone()).or_insert(0);
            let occurrence = *counter;
            *counter += 1;
            occurrence
        };
        tuples[index].occurrence = occurrence;
    }
    Ok(tuples)
}

/// Everything the cross-language check compares, per record in input
/// order. `identity.mjs` emits the identical JSON shape; the scope
/// audit fails on the first differing byte, naming the case and index.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct CaseIdentityReport {
    pub record_canonical: Vec<String>,
    pub identities: Vec<ExactIdentity>,
    pub identity_canonical: Vec<String>,
    pub identity_sha256: Vec<String>,
}

pub(crate) fn case_identity_report(
    fixture: &str,
    matrix_key: &str,
    oracle: &[GoldenDiag],
) -> ConformanceResult<CaseIdentityReport> {
    let identities = assign_case_identities(fixture, matrix_key, oracle)?;
    let record_canonical = oracle
        .iter()
        .map(|diag| {
            String::from_utf8(record_bytes(diag)).expect("canonical record bytes are UTF-8")
        })
        .collect();
    let identity_canonical = identities
        .iter()
        .map(|identity| {
            String::from_utf8(identity.canonical_bytes()).expect("canonical bytes are UTF-8")
        })
        .collect::<Vec<_>>();
    let identity_sha256 = identities.iter().map(ExactIdentity::sha256).collect();
    Ok(CaseIdentityReport {
        record_canonical,
        identities,
        identity_canonical,
        identity_sha256,
    })
}

// ---------------------------------------------------------------------------
// Required adversarial tests (measurement-integrity.md §7, A2 identity
// row) and the canonical-encoder canaries.
// ---------------------------------------------------------------------------

#[cfg(test)]
#[path = "../tests/unit/identity/tests.rs"]
mod tests;
